tests/runtimes/simple-keyvalue/src/methods.rs

```diff
@@ -350,6 +356,92 @@ impl Methods {
             .transpose()
             .map_err(|err| err.to_string())
     }
+
+    /// ElGamal encryption.
+    pub fn encrypt(ctx: &mut TxContext, args: Encrypt) -> Result<Option<String>, String> {
+        if ctx.is_check_only() {
+            return Ok(None);
+        }
+
+        // Derive key pair ID based on the given ID.
+        let key_pair_id = KeyPairId::from(Hash::digest_bytes(args.key_pair_id.as_bytes()).as_ref());
+
+        // Fetch public key.
+        let io_ctx = IoContext::create_child(&ctx.parent.core.io_ctx);
+        let result =
+            ctx.parent
+                .key_manager
+                .get_public_ephemeral_key(io_ctx, key_pair_id, args.epoch);
+        let long_term_pk = tokio::runtime::Handle::current()
+            .block_on(result)
+            .map_err(|err| err.to_string())?
+            .ok_or("public ephemeral key not available")?;
+
+        // Generate ephemeral key.
+        let mut rng = OsRng {};
+        let ephemeral_sk = x25519_dalek::StaticSecret::new(&mut rng);
+        let ephemeral_pk = x25519_dalek::PublicKey::from(&ephemeral_sk);
```

Comment:
This happens in a replicated manner across nodes and as such it will cause discrepancies if replication factor > 1 is used. Any randomness must be deterministic.

The only reason the E2E test passes is:
* In this test, the primary replication group size is 2 and the backup replication group size is 1. So even if this causes a discrepancy it is resolved by evaluating in the backup group which only has one member.
* The E2E test does not check whether any issues were detected. It needs to call `sc.Net.CheckLogWatchers()` at the end of the scenario.

Reply:
You are right. Now all makes sense as I was getting the following errors if I extended the client.
`"log assertion failed: round failure detected"`
`"log assertion failed: execution discrepancy detected"`